File: sol/solution/0001-0100/0016. 3Sum Closest/Solution.ex
defmodule Solution do
  @spec three_sum_closest(nums :: [integer], target :: integer) :: integer
  def three_sum_closest(nums, target) do
    nums = List.to_tuple(Enum.sort(nums))
    len = tuple_size(nums)
    cond do
      target > 3000 -> elem(nums, len - 1) + elem(nums, len - 2) + elem(nums, len - 3)
      target < -3000 -> elem(nums, 0) + elem(nums, 1) + elem(nums, 2)
      target > 0 -> do_three_sum_closest(nums, 0, len, target, -3000)
      true -> do_three_sum_closest(nums, 0, len, target, 3000)
    end
  end
  defp do_three_sum_closest(_nums, _i, _len, target, target), do: target
  defp do_three_sum_closest(_nums, i, len, _target, closest) when i > len - 3, do: closest
  defp do_three_sum_closest(nums, i, len, target, closest) do
    if i == 0 or elem(nums, i - 1) != elem(nums, i) do
      closest = find_closest(nums, i, i + 1, len - 1, target, closest)
      do_three_sum_closest(nums, i + 1, len, target, closest)
    else
      do_three_sum_closest(nums, i + 1, len, target, closest)
    end
  end
  defp find_closest(_nums, _i, l, r, _target, closest) when l >= r, do: closest
  defp find_closest(nums, i, l, r, target, closest) do
    val_i = elem(nums, i)
    val_l = elem(nums, l)
    val_r = elem(nums, r)
    sum = val_i + val_l + val_r
    cond do
      sum == target ->
        sum
      target > sum ->
        new_closest = update_closest(target, closest, sum)
        new_l = grow_from_left(nums, l + 1, r, val_l)
        find_closest(nums, i, new_l, r, target, new_closest)
      true ->
        new_closest = update_closest(target, closest, sum)
        new_r = shrink_from_right(nums, l, r - 1, val_r)
        find_closest(nums, i, l, new_r, target, new_closest)
    end
  end
  defp update_closest(target, current, new) do
    if abs(target - new) < abs(target - current) do
      new
    else
      current
    end
  end
  defp shrink_from_right(nums, l, r, prev_val_r) do
    cond do
      r <= l or elem(nums, r) != prev_val_r -> r
      true -> shrink_from_right(nums, l, r - 1, prev_val_r)
    end
  end
  defp grow_from_left(nums, l, r, prev_val_l) do
    cond do
      l >= r or elem(nums, l) != prev_val_l -> l
      true -> grow_from_left(nums, l + 1, r, prev_val_l)
    end
  end
end

File: sol/solution/0001-0100/0016. 3Sum Closest/Solution.erl
-spec three_sum_closest(Nums :: [integer()], Target :: integer()) -> integer().
three_sum_closest(Nums, Target) ->
    % I -> nums = [-1,2,1,-4], target = 1
    % O -> 2 (-1 + 2 + 1 = 2)
    Sorted = lists:sort(Nums),
    Array = array:from_list(Sorted),
    % [-4,-1,1,2]
    {_, Res} =
        array:foldl(fun (_I, V, {V, _} = Acc) ->
                            % skip duplicate value (vs prev)
                            Acc;
                        (I, V, {_, Prev}) ->
                            % 2 pointer movement
                            P1 = I + 1,
                            P2 = array:size(Array) - 1,
                            ClosestSum = find(Target, V, P1, P2, Array, Prev),
                            {V, closest(ClosestSum, Prev, Target)}
                    end,
                    {undefined, undefined},
                    Array),
    Res.
% returns closest sum to target for current E
find(_Target, _E, P1, P2, _Array, Res) when P1 >= P2 ->
    % finished traversing for E
    Res;
find(Target, E, P1, P2, Array, PrevSum) when P1 < P2 ->
    E1 = array:get(P1, Array),
    E2 = array:get(P2, Array),
    Sum = E + E1 + E2,
    % return triplet or move pointers
    case Sum of
        _ when Sum == Target ->
            % get the next pointers that skip past duplicates
            Next_P1 = pointer(P1, 1, array:size(Array) - 1, Array, E1),
            Next_P2 = pointer(P2, -1, P1, Array, E2),
            % return closest sum
            find(Target, E, Next_P1, Next_P2, Array, closest(Sum, PrevSum, Target));
        _ when Sum < Target ->
            find(Target, E, P1 + 1, P2, Array, closest(Sum, PrevSum, Target));
        _ when Sum > Target ->
            find(Target, E, P1, P2 - 1, Array, closest(Sum, PrevSum, Target))
    end.
closest(Sum, undefined, _Target) ->
    Sum;
closest(Sum, PrevSum, Target) ->
    A = abs(Target - Sum),
    B = abs(Target - PrevSum),
    R = min(A, B),
    case R of
        X when X == A ->
            Sum;
        Y when Y == B ->
            PrevSum
    end.
pointer(P, 1 = Amnt, Lim, Array, Prev) ->
    V = array:get(P, Array),
    case V of
        _ when V == Prev, P < Lim ->
            pointer(P + Amnt, Amnt, Lim, Array, Prev);
        _ ->
            P
    end;
pointer(P, -1 = Amnt, Lim, Array, Prev) ->
    V = array:get(P, Array),
    case V of
        _ when V == Prev, P > Lim ->
            pointer(P + Amnt, Amnt, Lim, Array, Prev);
        _ ->
            P
    end.